src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.cpp

```diff
@@ -0,0 +1,145 @@
+bool operator==(
+    const ScalarWaveGr<LocalScalarFieldData, LocalBackgroundData>& lhs,
+    const ScalarWaveGr<LocalScalarFieldData, LocalBackgroundData>&
+        rhs) noexcept {
+  return lhs.background_gr_data_ == rhs.background_gr_data_;
+}
+
+template <typename ScalarFieldData, typename BackgroundData>
+bool operator!=(
+    const ScalarWaveGr<ScalarFieldData, BackgroundData>& lhs,
+    const ScalarWaveGr<ScalarFieldData, BackgroundData>& rhs) noexcept {
+  return not(lhs == rhs);
+}
+
+// Generate instantiations
+#define STYPE(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define BTYPE(data) BOOST_PP_TUPLE_ELEM(1, data)
+
+#define INSTANTIATE(_, data)                                       \
+  template class ScalarWaveGr<STYPE(data), BTYPE(data)>;           \
+  template bool operator==(                                        \
+      const ScalarWaveGr<STYPE(data), BTYPE(data)>& lhs,           \
+      const ScalarWaveGr<STYPE(data), BTYPE(data)>& rhs) noexcept; \
+  template bool operator!=(                                        \
+      const ScalarWaveGr<STYPE(data), BTYPE(data)>& lhs,           \
+      const ScalarWaveGr<STYPE(data), BTYPE(data)>& rhs) noexcept;
+
+#define INSTANTIATE_CONSTRUCTOR1(_, data)                        \
+  template ScalarWaveGr<STYPE(data), BTYPE(data)>::ScalarWaveGr( \
+      const double mass,                                         \
+      const std::array<double, volume_dim> dimensionless_spin,   \
+      const std::array<double, volume_dim> center,               \
+      std::unique_ptr<MathFunction<1>> profile,                  \
+      const OptionContext& context) noexcept;
+
+#define INSTANTIATE_CONSTRUCTOR2(_, data)                        \
+  template ScalarWaveGr<STYPE(data), BTYPE(data)>::ScalarWaveGr( \
+      const double mass,                                         \
+      const std::array<double, volume_dim> dimensionless_spin,   \
+      const std::array<double, volume_dim> center,               \
+      std::array<double, volume_dim> wave_vector,                \
+      std::array<double, volume_dim> wave_center,                \
+      std::unique_ptr<MathFunction<1>> profile,                  \
+      const OptionContext& context) noexcept;
+
+GENERATE_INSTANTIATIONS(INSTANTIATE,
+                        (ScalarWave::Solutions::RegularSphericalWave,
+                         ScalarWave::Solutions::PlaneWave<3>),
+                        (gr::Solutions::KerrSchild))
+GENERATE_INSTANTIATIONS(INSTANTIATE_CONSTRUCTOR1,
+                        (ScalarWave::Solutions::RegularSphericalWave),
+                        (gr::Solutions::KerrSchild))
+GENERATE_INSTANTIATIONS(INSTANTIATE_CONSTRUCTOR2,
+                        (ScalarWave::Solutions::PlaneWave<3>),
+                        (gr::Solutions::KerrSchild))
+
+#undef INSTANTIATE
+#undef INSTANTIATE_CONSTRUCTOR1
+#undef INSTANTIATE_CONSTRUCTOR2
+#undef STYPE
+
+}  // namespace CurvedScalarWave::AnalyticData
+/// \endcond
```

Comment:
turn this file into a `.tpp` file dropping the instantiations.  Then for each combination of ScalarWave/Background create a simple `.cpp` and `.hpp` pair where the `.cpp` file has the instantiation and the `.hpp` file just has includes of the specific flat-space solution, background solution and the `tpp` file

Reply:
updated